File: curator/features/profiles.py
"""Inspectable, missing-aware performer profiles and similarity."""

from __future__ import annotations

import math
from dataclasses import dataclass, field


@dataclass(frozen=True)
class ProfileValue:
    value: float
    confidence: float


@dataclass(frozen=True)
class PerformerProfile:
    performer_id: str
    blocks: dict[str, dict[str, ProfileValue]]
    norms: dict[str, float] = field(init=False, repr=False, compare=False)
    keys: dict[str, frozenset[str]] = field(init=False, repr=False, compare=False)

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "norms",
            {
                block: math.sqrt(sum(item.value**2 for item in values.values()))
                for block, values in self.blocks.items()
                if block not in NUMERIC_BLOCKS
            },
        )
        # Frozen key sets avoid rebuilding a set from each block dict on every
        # similarity call; the builder compares every profile against every known
        # profile, so this is the difference between tens of millions of conversions
        # and a single hash each.
        object.__setattr__(
            self, "keys", {block: frozenset(values) for block, values in self.blocks.items()}
        )


@dataclass(frozen=True)
class SimilarityResult:
    similarity: float
    block_similarities: dict[str, float]
    block_weights: dict[str, float]


NUMERIC_SCALES = {
    "height_cm": 12.0,
    "weight_kg": 15.0,
    "band_inches": 6.0,
    "cup_index": 2.0,
    "waist_inches": 7.0,
    "hip_inches": 8.0,
    "waist_to_hip": 0.12,
    "waist_to_height": 0.10,
    "hip_to_height": 0.12,
    "age_recording": 8.0,
}
NUMERIC_BLOCKS = {"measurements", "height", "age"}


def _cosine(
    left: dict[str, ProfileValue],
    right: dict[str, ProfileValue],
    left_norm: float,
    right_norm: float,
    left_keys: frozenset[str] | None = None,
    right_keys: frozenset[str] | None = None,
) -> float | None:
    # Precomputed frozensets keep the same iteration order as freshly built sets
    # (both derive from the same dict insertion order), so dot-product accumulation
    # stays bit-identical to the previous set(left) & set(right) form.
    shared = (left_keys or frozenset(left)) & (right_keys or frozenset(right))
    if not shared:
        return 0.0
    dot = 0.0
    confidence_sum = 0.0
    for key in shared:
        left_value = left[key]
        right_value = right[key]
        dot += left_value.value * right_value.value
        right_confidence = right_value.confidence
        left_confidence = left_value.confidence
        confidence_sum += (
            left_confidence if left_confidence < right_confidence else right_confidence
        )
    if left_norm == 0 or right_norm == 0:
        return None
    confidence = confidence_sum / len(shared)
    return max(0.0, min(1.0, dot / (left_norm * right_norm) * confidence))


def _numeric(
    left: dict[str, ProfileValue],
    right: dict[str, ProfileValue],
    left_keys: frozenset[str] | None = None,
    right_keys: frozenset[str] | None = None,
) -> float | None:
    shared = (left_keys or frozenset(left)) & (right_keys or frozenset(right))
    if not shared:
        return None
    values = []
    for key in sorted(shared):
        scale = NUMERIC_SCALES.get(key, 1.0)
        left_value = left[key]
        right_value = right[key]
        closeness = math.exp(-abs(left_value.value - right_value.value) / scale)
        right_confidence = right_value.confidence
        left_confidence = left_value.confidence
        values.append(
            closeness
            * (left_confidence if left_confidence < right_confidence else right_confidence)
        )
    return sum(values) / len(values)


def block_similarity(
    left: PerformerProfile,
    right: PerformerProfile,
    block: str,
) -> float | None:
    """Compare one shared block, or return None when it carries no usable evidence."""
    if block not in left.blocks or block not in right.blocks:
        return None
    if block in NUMERIC_BLOCKS:
        return _numeric(
            left.blocks[block], right.blocks[block], left.keys[block], right.keys[block]
        )
    return _cosine(
        left.blocks[block],
        right.blocks[block],
        left.norms[block],
        right.norms[block],
        left.keys[block],
        right.keys[block],
    )


def block_similarities(
    left: PerformerProfile,
    right: PerformerProfile,
    block_weights: dict[str, float],
) -> tuple[dict[str, float], dict[str, float]]:
    """Per-block similarities and the weights they were measured with."""
    similarities: dict[str, float] = {}
    used_weights: dict[str, float] = {}
    for block in sorted(set(left.blocks) & set(right.blocks)):
        weight = block_weights.get(block, 0.0)
        if weight <= 0:
            continue
        similarity = block_similarity(left, right, block)
        if similarity is None:
            continue
        similarities[block] = similarity
        used_weights[block] = weight
    return similarities, used_weights


def similarity_penalty(left: PerformerProfile, right: PerformerProfile) -> float:
    """Scale contradicting body evidence down; depends on no block that varies by scene."""
    penalty = 1.0
    left_cup = left.blocks.get("measurements", {}).get("cup_index")
    right_cup = right.blocks.get("measurements", {}).get("cup_index")
    if left_cup and right_cup:
        penalty *= math.exp(-0.18 * max(0.0, abs(left_cup.value - right_cup.value) - 1))
    left_aug = set(left.blocks.get("augmentation", {}))
    right_aug = set(right.blocks.get("augmentation", {}))
    if left_aug and right_aug and not left_aug & right_aug:
        penalty *= 0.65
    return penalty


def combine_similarities(
    left: PerformerProfile,
    right: PerformerProfile,
    similarities: dict[str, float],
    used_weights: dict[str, float],
) -> SimilarityResult:
    """Weight measured blocks into one score, so callers can reuse per-block work."""
    denominator = sum(used_weights.values())
    total = (
        sum(similarities[block] * used_weights[block] for block in similarities) / denominator
        if denominator
        else 0.0
    )
    return SimilarityResult(total * similarity_penalty(left, right), similarities, used_weights)


def performer_similarity(
    left: PerformerProfile,
    right: PerformerProfile,
    block_weights: dict[str, float],
) -> SimilarityResult:
    similarities, used_weights = block_similarities(left, right, block_weights)
    return combine_similarities(left, right, similarities, used_weights)

File: core/profiles.go
// Performer profile similarity — a port of curator/features/profiles.py.
//
// Block math mirrors the Python exactly: numeric blocks (measurements,
// height, age) accumulate over SORTED shared keys; cosine blocks accumulate
// over set-intersection order, which in CPython is hash order. For the
// shipped feature builder every cosine block carries at most one value, so
// the accumulation order is deterministic in practice; the differential
// harness covers this.
package main

import (
	"math"
	"sort"
)

// numericScales mirrors profiles.NUMERIC_SCALES.
var numericScales = map[string]float64{
	"height_cm":       12.0,
	"weight_kg":       15.0,
	"band_inches":     6.0,
	"cup_index":       2.0,
	"waist_inches":    7.0,
	"hip_inches":      8.0,
	"waist_to_hip":    0.12,
	"waist_to_height": 0.10,
	"hip_to_height":   0.12,
	"age_recording":   8.0,
}

// numericBlocks mirrors profiles.NUMERIC_BLOCKS.
var numericBlocks = map[string]bool{"measurements": true, "height": true, "age": true}

// performerBlockWeights mirrors FeatureConfig.performer_block_weights in
// tuple order (dict(block_weights) preserves it).
var performerBlockWeights = []struct {
	block  string
	weight float64
}{
	{"content", 1.0},
	{"measurements", 1.0},
	{"augmentation", 0.9},
	{"ethnicity", 0.8},
	{"height", 0.7},
	{"age", 0.6},
	{"hair", 0.45},
	{"tattoos", 0.35},
	{"piercings", 0.25},
	{"eyes", 0.1},
}

// blockSimilarity mirrors profiles.block_similarity.
func blockSimilarity(left, right *performerProfile, block string) (float64, bool) {
	leftBlock, leftOK := left.blocks[block]
	rightBlock, rightOK := right.blocks[block]
	if !leftOK || !rightOK {
		return 0, false
	}
	if numericBlocks[block] {
		return numericSimilarity(leftBlock, rightBlock, left.keys[block], right.keys[block])
	}
	return cosineSimilarity(leftBlock, rightBlock, left.norms[block], right.norms[block], left.keys[block], right.keys[block])
}

// numericSimilarity mirrors profiles._numeric: shared keys sorted, mean of
// exp(-|diff|/scale) * min confidence.
func numericSimilarity(left, right map[string]profileValue, leftKeys, rightKeys map[string]bool) (float64, bool) {
	shared := intersectKeys(leftKeys, rightKeys)
	if len(shared) == 0 {
		return 0, false
	}
	sort.Strings(shared)
	sum := 0.0
	for _, key := range shared {
		scale := numericScales[key]
		if scale == 0 {
			scale = 1.0
		}
		leftValue := left[key]
		rightValue := right[key]
		closeness := pyExp(-math.Abs(leftValue.value-rightValue.value) / scale)
		confidence := leftValue.confidence
		if rightValue.confidence < confidence {
			confidence = rightValue.confidence
		}
		sum += closeness * confidence
	}
	return sum / float64(len(shared)), true
}

// cosineSimilarity mirrors profiles._cosine: dot over shared keys, scaled by
// the mean pairwise-min confidence, clamped to [0, 1].
func cosineSimilarity(left, right map[string]profileValue, leftNorm, rightNorm float64, leftKeys, rightKeys map[string]bool) (float64, bool) {
	shared := intersectKeys(leftKeys, rightKeys)
	if len(shared) == 0 {
		return 0.0, true
	}
	dot := 0.0
	confidenceSum := 0.0
	for _, key := range shared {
		leftValue := left[key]
		rightValue := right[key]
		dot += leftValue.value * rightValue.value
		confidence := leftValue.confidence
		if rightValue.confidence < confidence {
			confidence = rightValue.confidence
		}
		confidenceSum += confidence
	}
	if leftNorm == 0 || rightNorm == 0 {
		return 0, false
	}
	confidence := confidenceSum / float64(len(shared))
	value := dot / (leftNorm * rightNorm) * confidence
	if value < 0 {
		value = 0
	}
	if value > 1 {
		value = 1
	}
	return value, true
}

// intersectKeys returns the keys present in both sets. Iteration order only
// matters for the cosine accumulation, and the shipped cosine blocks hold at
// most one shared key, so order is deterministic in practice.
func intersectKeys(left, right map[string]bool) []string {
	if len(left) > len(right) {
		left, right = right, left
	}
	shared := make([]string, 0, len(left))
	for key := range left {
		if right[key] {
			shared = append(shared, key)
		}
	}
	return shared
}

// similarityPenalty mirrors profiles.similarity_penalty.
func similarityPenalty(left, right *performerProfile) float64 {
	penalty := 1.0
	leftCup, leftOK := profileValueAt(left, "measurements", "cup_index")
	rightCup, rightOK := profileValueAt(right, "measurements", "cup_index")
	if leftOK && rightOK {
		penalty *= pyExp(-0.18 * mathMax(0.0, math.Abs(leftCup.value-rightCup.value)-1))
	}
	leftAug, leftOK := left.blocks["augmentation"]
	rightAug, rightOK := right.blocks["augmentation"]
	if leftOK && rightOK && len(leftAug) > 0 && len(rightAug) > 0 && !keysOverlap(leftAug, rightAug) {
		penalty *= 0.65
	}
	return penalty
}

func profileValueAt(p *performerProfile, block, key string) (profileValue, bool) {
	values, ok := p.blocks[block]
	if !ok {
		return profileValue{}, false
	}
	v, ok := values[key]
	return v, ok
}

func keysOverlap(a, b map[string]profileValue) bool {
	for key := range a {
		if _, ok := b[key]; ok {
			return true
		}
	}
	return false
}

// blockSimilaritiesAll mirrors profiles.block_similarities: per-block
// similarities and the weights they were measured with, over the sorted
// shared blocks (the Python dict insertion order), plus the ordered block
// names and weights for deterministic accumulation.
func blockSimilaritiesAll(left, right *performerProfile, blockWeights map[string]float64) (map[string]float64, map[string]float64, []string, []float64) {
	blocks := make([]string, 0)
	for block := range left.blocks {
		if _, ok := right.blocks[block]; ok {
			blocks = append(blocks, block)
		}
	}
	sort.Strings(blocks)
	similarities := make(map[string]float64)
	usedWeights := make(map[string]float64)
	var ordered []string
	var weights []float64
	for _, block := range blocks {
		weight := blockWeights[block]
		if weight <= 0 {
			continue
		}
		similarity, ok := blockSimilarity(left, right, block)
		if !ok {
			continue
		}
		similarities[block] = similarity
		usedWeights[block] = weight
		ordered = append(ordered, block)
		weights = append(weights, weight)
	}
	return similarities, usedWeights, ordered, weights
}

// performerSimilarity mirrors profiles.performer_similarity; returns the
// combined similarity plus the per-block similarities and used weights.
func performerSimilarity(left, right *performerProfile, blockWeights map[string]float64) (float64, map[string]float64, map[string]float64) {
	similarities, usedWeights, ordered, weights := blockSimilaritiesAll(left, right, blockWeights)
	denominator := neumaierSum(weights)
	total := 0.0
	if denominator > 0 {
		products := make([]float64, 0, len(ordered))
		for _, block := range ordered {
			products = append(products, similarities[block]*usedWeights[block])
		}
		total = neumaierSum(products) / denominator
	}
	return total * similarityPenalty(left, right), similarities, usedWeights
}

// performerBlockWeightsMap mirrors DEFAULT_CONFIG.feature.performer_block_weights
// as a lookup map.
func performerBlockWeightsMap() map[string]float64 {
	weights := map[string]float64{}
	for _, entry := range performerBlockWeights {
		weights[entry.block] = entry.weight
	}
	return weights
}
